sycl/doc/Assert.md

```diff
@@ -0,0 +1,426 @@
+It guarantees assertion failure notification delivery to the host regardless of
+kernel behavior which hit the assertion.
+
+The Native Device Compiler is responsible for providing implementation of
+`__devicelib_assert_fail` which completely hides details of communication
+between the device code and the Low-Level Runtime from the SYCL device compiler
+and runtime. The Low-Level Runtime is responsible for:
+ - detecting if assert failure took place;
+ - flushing assert message to `stderr` on host.
+
+When detected, Low-level Runtime reports assert failure to DPCPP Runtime
+via events objects. To achieve this, information about assert failure should be
+propagated from device-side to SYCL Runtime. This should be performed via calls
+to `piEventGetInfo`. This Plugin Interface call "lowers" to `clGetEventInfo` for
+OpenCL backend and `zeEventQueryStatus` for Level-Zero backend.
+
+Refer to [OpenCL](extensions/Assert/opencl.md) and [Level-Zero](extensions/Assert/level-zero.md)
+extensions.
+
+The following sequence of events describes how user code gets notified:
+ - Device side:
+   1. Assert fails in device-code in kernel
+      // It's not defined if GPU thread stops execution
+      // Other GPU threads are left untouched
+   2. Specialized version of `__devicelib_assert_fail` is called
+   3. Device immediately signals to host (Low-Level Runtime)
+ - Host side:
+   1. The assert failure gets detected by Low-Level Runtime
+   2. Low-Level Runtime prints assert failure message to `stderr`
+   3. Low-Level Runtime calls `abort()`
+
+
+### Fallback approach
+
+If Device-side Runtime doesn't support `__devicelib_assert_fail` then a buffer
+based approach comes in place. The approach doesn't require any support from
+Device-side Runtime and Native Device Compiler. Neither it does from Low-level
+Runtime.
+
+Within this approach, a dedicated assert buffer is allocated and implicit kernel
+argument is introduced. The argument is an accessor with `discard_read_write`
+or `discard_write` access mode. Accessor metadata is stored to program scope
+variable. This allows to refer to the accessor without modifying each and every
+user's function. Fallback implementation of `__devicelib_assert_fail` restores
+accessor metadata from program scope variable and writes assert information to
+the assert buffer. Atomic operations are used in order to not overwrite existing
+information.
+
+DPCPP Runtime checks contents of the assert buffer for assert failure flag after
+kernel finishes.
+
+The following sequence of events describes how user code gets notified:
+ - Device side:
+   1. Assert fails in device-code in kernel
+   2. Fallback version of `__devicelib_assert_fail` is called
+   3. Assert information is stored into assert buffer
+   4. Kernel continues running
+ - Host side:
+   1. A distinct thread is launched no later than the point of enqueue of the of
+      kernel with assertions
```

Comment:
+1 Will requirements be met if we do checks on blocking calls?

Reply:
I think we should be concerned about a thread spinning inside the DPC++ runtime whenever a kernel is active.  This wastes a CPU in the case where there is other useful work going on in the CPU.  Even if there is no other useful CPU work, it wastes power by keeping the CPU active, and this could be particularly important in a large HPC or datacenter system.

One solution is to check the buffer only when the DPC++ runtime normally to see if a kernel completes.  Does this only happen when the application makes a blocking call?  I think the infinite spin loop example that @s-kanaev shows above is probably not a realistic example.  Application code will eventually make a blocking call to know when its kernel completes.

Another solution is to have the internal spin loop, but ensure that it is blocked most of the time.  For example, it can call sleep() and only check the buffer very infrequently.

Reply:
>  Does this only happen when the application makes a blocking call?

he check can be performed with using of host-task also.

Reply:
I think you could also just do this check every time the DPC++ runtime makes a call to the plugin, as a sort of wrapper at the beginning of the plug-in call.